### PR TITLE
fix(providers): redact embedded credentials from baseURL in debug logs

### PR DESCRIPTION
**Pull request by murdore**
## What & why

Follow-up to the provider-migration series (releases 9.68.11–9.68.16). This addresses the CodeRabbit "Major" findings raised on #1050 (groq) and #1053 (cohere): a user-configurable `baseURL` (supplied via credentials or env) can carry inline `user:pass@host` credentials, and the provider **constructor debug logs** plus the **"not reachable" / "endpoint not available" error messages** previously emitted that `baseURL` verbatim — so credentials could land in debug output.

Rather than patch one provider at a time, this applies the fix consistently across the whole OpenAI-compatible cohort and promotes the existing `llamaCpp` helper to a shared util.

## Changes

- **Extract** `redactUrlCredentials(url)` out of `llamaCpp` into the shared `src/lib/utils/logSanitize.ts` util (turns `https://user:pass@host` into `https://***@host`; no-op for credential-free URLs). `llamaCpp` now imports it instead of defining its own copy.
- **Wrap `baseURL` in the constructor debug log** of: `groq`, `cohere`, `deepseek`, `perplexity`, `togetherAi`, `openAI`, `nvidiaNim`, `cloudflare`, `openaiCompatible`, `litellm`, `huggingFace`, `xai`, `lmStudio`, `fireworks`.
- **Wrap the user-facing endpoint error message** in `lmStudio` ("server not reachable at …") and `openaiCompatible` ("endpoint not available … OPENAI_COMPATIBLE_BASE_URL: …").

## Deliberately untouched

- The public `getConfiguration()` contract returns the **raw** `baseURL` (callers depend on the real value) — verified for `cohere`, `nvidiaNim`, `lmStudio`, and the base class.
- All request-URL construction (`${stripTrailingSlash(this.config.baseURL)}/...`, etc.) keeps the **raw** `baseURL` — redacting it would break requests.
- Image/media providers (`ideogram`, `recraft`, `replicate`, `voyage`, `stability`, `jina`) use a **fixed provider `baseURL`** with no user-supplied credentials, so there is nothing to redact — out of scope by design.

## Validation

- `tsc --noEmit --strict` — pass
- lint + format — pass (0 errors)
- `pnpm run build` + publint — pass ("All good!")
- stream-span native-base audit — 106/106 pass
- `validate:all` (Gitleaks) — no secrets detected

Single commit, per the repo's single-commit policy.

## Summary by CodeRabbit

* **Bug Fixes**
  * Added automatic credential redaction across all AI provider integrations, ensuring sensitive URL information no longer appears in debug logs or error messages.

### CODE REVIEW

File: src/lib/utils/logSanitize.ts [hunk 1]
```diff
@@ -100,6 +100,21 @@ export function sanitizeForLog(text: string, maxLen = 500): string {
...
+ * Turns `https://user:secret@host/path` into `https://***@host/path` while
+ * leaving credential-free URLs untouched, so the host/port/path stay useful
+ * for diagnostics. Use this for any `baseURL`/endpoint a caller may have
+ * supplied with inline credentials.
+ *
+ * @param url - The URL (or URL-shaped string) to redact.
+ */
+export function redactUrlCredentials(url: string): string {
+  return url.replace(/\/\/[^/@]+@/, "//***@");
+}
```
Comment (Tara-ag) [inline src/lib/utils/logSanitize.ts:118]: ⚠️ **MAJOR**: The regex `/\/\/[^/@]+@/` only replaces the **first** occurrence of embedded credentials. If a URL contains multiple credential sections (e.g., proxy chains or callback URLs with embedded auth), subsequent credentials would still leak into logs.

**Example:**
```javascript
// Current behavior (partial redaction):
"https://user:pass@host/callback?next=https://other:secret@elsewhere"
// → "https://***@host/callback?next=https://other:secret@elsewhere"

// Expected (full redaction):
"https://***@host/callback?next=https://***@elsewhere"
```

**Suggested fix:**
```typescript
export function redactUrlCredentials(url: string): string {
  return url.replace(/\/\/[^/@]+@/g, "//***@");
}
```

Note: The `g` (global) flag ensures all occurrences are redacted.

Comment (Tara-ag) [inline src/lib/utils/logSanitize.ts:118]: 💡 **MINOR**: This new security-critical function lacks test coverage in the existing log-sanitization verification suite (`test/continuous-test-suite-log-sanitize.ts`).

The existing test suite covers `sanitizeForLog`, `sanitizeRecord`, and `sanitizeHeaders` extensively. Adding regression tests for `redactUrlCredentials` would help prevent accidental reintroductions of credential leakage.

**Suggested test cases to add:**
```typescript
// Basic credential redaction
expect(redactUrlCredentials("https://user:pass@host/path")).toBe("https://***@host/path");

// Multiple credential sections
expect(redactUrlCredentials("https://a:b@host?next=https://c:d@other"))
  .toBe("https://***@host?next=https://***@other");

// Credential-free URLs unchanged
expect(redactUrlCredentials("https://host/path")).toBe("https://host/path");

// Edge cases
expect(redactUrlCredentials("")).toBe("");
expect(redactUrlCredentials("not-a-url")).toBe("not-a-url");
```

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Review Summary

**Files reviewed:** 16 changed files
**New issues found:** 2 (1 MAJOR, 1 MINOR)

### 🔒 Blocking Issues (must fix)

**MAJOR: Non-global regex in `redactUrlCredentials`** (`src/lib/utils/logSanitize.ts:116`)

The regex `/\/\/[^/@]+@/` only replaces the **first** occurrence of embedded credentials. If a URL contains multiple credential sections (e.g., proxy chains or callback URLs with embedded auth), subsequent credentials would still leak into logs.

**Example of the bug:**
```javascript
// Current behavior (partial redaction - second credential leaked):
"https://user:pass@host/callback?next=https://other:secret@elsewhere"
// → "https://***@host/callback?next=https://other:secret@elsewhere"

// Expected (full redaction):
"https://***@host/callback?next=https://***@elsewhere"
```

**Fix:** Add the global flag `g` to the regex:
```typescript
return url.replace(/\/\/[^/@]+@/g, "//***@");
```

### 💡 Minor Issues

**MINOR: Missing test coverage** (`src/lib/utils/logSanitize.ts`)

The new security-critical `redactUrlCredentials` function lacks test coverage in the existing log-sanitization verification suite (`test/continuous-test-suite-log-sanitize.ts`). Consider adding regression tests for:
- Basic credential redaction
- Multiple credential sections (edge case)
- Credential-free URLs (unchanged)
- Empty/invalid inputs

### ✅ What's Done Well

- Consistent application across all 16 OpenAI-compatible providers
- Proper deduplication by extracting the helper from `llamaCpp` to shared utils
- Correctly preserves raw `baseURL` in `getConfiguration()` (public API contract)
- Correctly preserves raw `baseURL` in request URL construction
- Error messages in `lmStudio` and `openaiCompatible` properly redacted

### Next Steps

1. Fix the regex to use global replace (`/g` flag)
2. (Optional) Add test coverage for `redactUrlCredentials`
3. Re-run validation suite (`tsc`, lint, tests)

Once the MAJOR issue is addressed, this PR is ready to merge.

File: src/lib/utils/logSanitize.ts (hunk 1 above)
Comment (murdore) [reply src/lib/utils/logSanitize.ts:118]: Fixed in 4bd0983c. The regex now carries the global flag (//[^/@]+@/g), so every //...@ authority in the string is redacted, not just the first. Covered by a multi-credential case in the log-sanitize suite: "https://u1:p1@host-a/path?next=https://u2:p2@host-b/cb" -> both credential pairs stripped.

[the remark this answers is not in the document]
Comment (murdore) [reply src/lib/utils/logSanitize.ts:118]: Fixed in 4bd0983c. The regex now carries the global flag (//[^/@]+@/g), so every //...@ authority in the string is redacted, not just the first. Covered by a multi-credential case in the log-sanitize suite: "https://u1:p1@host-a/path?next=https://u2:p2@host-b/cb" -> both credential pairs stripped.

Comment (murdore) [reply src/lib/utils/logSanitize.ts:118]: Added in 4bd0983c. Section I of test/continuous-test-suite-log-sanitize.ts now covers redactUrlCredentials: credential stripping across https/redis/percent-encoded URLs, global multi-occurrence redaction, and credential-free no-op cases (plain baseURL, localhost, path-only, empty string). Suite passes 49/49.

[the remark this answers is not in the document]
Comment (murdore) [reply src/lib/utils/logSanitize.ts:118]: Added in 4bd0983c. Section I of test/continuous-test-suite-log-sanitize.ts now covers redactUrlCredentials: credential stripping across https/redis/percent-encoded URLs, global multi-occurrence redaction, and credential-free no-op cases (plain baseURL, localhost, path-only, empty string). Suite passes 49/49.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Files reviewed:** 17  
**New issues found:** 0  
**Existing comments addressed:** 4 (all resolved by author)

### Assessment

This PR successfully addresses the security concern of credential leakage in debug logs and error messages. The implementation is clean, consistent, and follows the project's architectural patterns.

#### What's Good

1. **Security fix is correct**: The `redactUrlCredentials()` function properly strips `user:pass@` credentials from URLs using a global regex (`/\/\/[^/@]+@/g`), ensuring all occurrences are redacted, not just the first.

2. **Comprehensive coverage**: All 14 OpenAI-compatible providers now wrap `baseURL` in their constructor debug logs. The `lmStudio` and `openaiCompatible` providers also redact credentials in user-facing error messages.

3. **Proper boundaries respected**: 
   - Raw `baseURL` is preserved in `getConfiguration()` returns (public API contract)
   - Raw `baseURL` is used for actual request construction (no breakage)
   - Only debug logs and error messages are sanitized

4. **Good refactoring**: The `llamaCpp` provider's local `redactUrlCredentials` function was correctly extracted to the shared utility.

5. **Test coverage**: Section I in `continuous-test-suite-log-sanitize.ts` covers positive cases (HTTPS, Redis, percent-encoded credentials), multi-credential global redaction, and negative cases (credential-free URLs, empty strings).

#### Minor Observation (Non-blocking)

There's a minor inconsistency in how `baseURL` is handled across providers:
- Most providers use `.trim()` on `credentials?.baseURL?.trim()` before fallback
- `nvidiaNim.ts` and `lmStudio.ts` use `credentials?.baseURL ?? envVar` without trim

This is pre-existing behavior and not introduced by this PR. The PR maintains consistency with each provider's existing pattern.

### Verification

- ✅ No hardcoded secrets
- ✅ No breaking changes to public SDK API
- ✅ Follows CLAUDE.md Critical Rules
- ✅ All existing review comments have been addressed (global regex flag, test coverage)
- ✅ Type safety maintained
- ✅ Error handling preserved

**Recommendation:** Approve for merge.
